src/middlewared/middlewared/plugins/replication.py

```diff
@@ -1,108 +1,328 @@
+                'zfs.pool.query',
+                [('name', '=', dest_dataset_name.split('/')[0])]
+            )
+            if dest_pool:
+                sender = subprocess.Popen(
+                        ['zfs', 'send', '-p', snap_sorted_s[-1].name],
+                        stdout=subprocess.PIPE
+                    )
+            else:
+                await self.do_update(repl_task_id, {'repl_result': 'Destination pool does not exists'})
+                raise CallError('Destination pool does not exists')
+        else:
+
+            if token.value and repl_task['repl_resume']:
+                sender = subprocess.Popen(
+                    ['zfs', 'send', '-t', '-p', token.value, dest_dataset_name],
+                    stdout=subprocess.PIPE
+                )
+            elif token.value and not repl_task['repl_resume']:
+                subprocess.Popen(['zfs', 'recv', '-A', dest_dataset_name])
+
+            else:
+                dest_snapshots = src_dataset.snapshots
+                snap_sorted_r = sorted(dest_snapshots, key=lambda x: x.properties['createtxg'].rawvalue)
+                last_common_snapshot = find_latest_common_snap(snap_sorted_s, snap_sorted_r)
+
+                if last_common_snapshot.split('/')[-1] == snap_sorted_s[-1].name:
+                    await self.do_update(repl_task_id, {'repl_result': f'Destination dataset already has the newest snapshot'})
+                    return False
+                else:
+                    sender = subprocess.Popen(
+                        ['zfs', 'send', '-p', '-I', last_common_snapshot, snap_sorted_s[-1].name],
+                        stdout=subprocess.PIPE)
+
+        receiver = subprocess.Popen(
+            ['zfs', 'recv', '-s' if repl_task['repl_resume'] else '', repl_task['repl_zfs']],
+            stdin=sender.stdout
         )
 
-        return {
-            'ssh_port': ssh['ssh_tcpport'],
-            'ssh_hostkey': ssh_hostkey,
-        }
+        while True:
+            # Just for PoC
+            if sender.poll() is not None or receiver.poll() is not None:
+                receiver.terminate()
+                sender.terminate()
+                await self.do_update(repl_task_id, {'repl_result': f'Sending process finished with {sender.returncode} retruncode and receiving with {receiver.returncode}'})
+                break
+
+        if repl_task['repl_followdelete']:
+            snaps_to_delete = find_snaps_to_delete(snap_sorted_s, snap_sorted_r)
+            for snap in snaps_to_delete:
+                snap = zfs.get_snapshot(snap)
+                snap.delete(True)
+
+        return True
+
+    @job(process=True)
+    async def start_ssh_replication(self, job, repl_task_id):
+        zfs = libzfs.ZFS()
+
+        try:
+            repl_task = await self.middleware.call(
+                'replication.task.query',
+                [('id', '=', repl_task_id)],
+                {'get': True}
+            )
+        except IndexError:
```

Comment:
job with process=True is more of a hack.

middleware.call on a job process will not raise an IndexError, but a ClientException